src-java/stats-topology/stats-storm-topology/src/test/java/org/openkilda/wfm/topology/stats/StatsTopologyTest.java

```diff
@@ -426,6 +494,519 @@ public void flowStatsTest() {
+        validateHaFlowMeterStats(forward, haFlow.getForwardPath().getCookie(), SWITCH_ID_2,
+                METER_ID_HA_FLOW_PATH_FORWARD.getValue(), SUB_FLOW_ID_SHARED, true);
+
+
+        //reverse
+        FlowSegmentCookie reverseCookie = haFlow.getReversePath().getCookie();
+
+        MeterStatsEntry reverse1 = new MeterStatsEntry(METER_ID_REVERSE_SUB_PATH_1.getValue(), 400L, 500L);
+        sendStatsMessage(new MeterStatsData(SWITCH_ID_3, Collections.singletonList(reverse1)));
+        validateHaFlowMeterStats(reverse1, reverseCookie.toBuilder()
+                        .subType(FlowSegmentCookie.FlowSubType.HA_SUB_FLOW_1).build(),
+                SWITCH_ID_3, METER_ID_REVERSE_SUB_PATH_1.getValue(), SUB_FLOW_ID_1, false);
+
+        MeterStatsEntry reverse2 = new MeterStatsEntry(METER_ID_REVERSE_SUB_PATH_2.getValue(), 400L, 500L);
+        sendStatsMessage(new MeterStatsData(SWITCH_ID_4, Collections.singletonList(reverse2)));
+        validateHaFlowMeterStats(reverse2, reverseCookie.toBuilder()
+                        .subType(FlowSegmentCookie.FlowSubType.HA_SUB_FLOW_2).build(),
+                SWITCH_ID_4, METER_ID_REVERSE_SUB_PATH_2.getValue(), SUB_FLOW_ID_2, false);
+    }
+
+    /**
+     * 3 switch ha-flow flow-stats test.
+     * ''''''/----3
+     * 1----2
+     */
+    @Test
+    public void haFlowStatsTest3() {
+        HaFlow haFlow = createHaFlow3();
+
+        sendUpdateHaFlowPathInfo(haFlow);
+        //forward
+        FlowStatsEntry forwardIngress = new FlowStatsEntry(1, haFlow.getForwardPath().getCookie().getValue(),
+                150L, 300L, 10, 10);
+        sendStatsMessage(new FlowStatsData(SWITCH_ID_1, Collections.singletonList(forwardIngress)));
+        validateHaFlowStats(forwardIngress, haFlow.getForwardPath().getCookie(), SWITCH_ID_1, true,
+                false, false, SUB_FLOW_ID_SHARED);
+
+        FlowStatsEntry forwardYPoint = new FlowStatsEntry(1, haFlow.getForwardPath().getCookie().getValue(),
+                150L, 300L, 10, 10);
+        sendStatsMessage(new FlowStatsData(SWITCH_ID_2, Collections.singletonList(forwardYPoint)));
+        validateHaFlowStats(forwardYPoint, haFlow.getForwardPath().getCookie(), SWITCH_ID_2, false,
+                true, true, SUB_FLOW_ID_SHARED);
+
+        FlowStatsEntry forwardEgressPoint1 = new FlowStatsEntry(1, COOKIE_FORWARD.toBuilder()
+                .subType(FlowSegmentCookie.FlowSubType.HA_SUB_FLOW_1).build().getValue(), 150L,
+                300L, 10, 10);
+        sendStatsMessage(new FlowStatsData(SWITCH_ID_3, Collections.singletonList(forwardEgressPoint1)));
+        validateHaFlowStats(forwardEgressPoint1, haFlow.getForwardPath().getCookie().toBuilder()
+                        .subType(FlowSegmentCookie.FlowSubType.HA_SUB_FLOW_1).build(), SWITCH_ID_3, false,
+                true, false, SUB_FLOW_ID_1);
+
+        FlowStatsEntry forwardEgressPoint2 = new FlowStatsEntry(1, haFlow.getForwardPath().getCookie().getValue(),
```

Comment:
forwardEgressPoint2 and forwardYPoint is the same. So you can remove forwardEgressPoint2.

Also forwardEgressPoint2 is Y point too, but you set `yPoint` flam in validate method to `false`

Reply:
thank you, fixed